<!DOCTYPE html>

<html lang="en" data-content_root="./">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Example application of quipcell to human lung cell atlas &#8212; quipcell 0.8 documentation</title>
    <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="_static/alabaster.css?v=89af775e" />
    <link rel="stylesheet" type="text/css" href="_static/nbsphinx-code-cells.css" />
    <script src="_static/documentation_options.js?v=a0e24af7"></script>
    <script src="_static/doctools.js?v=9a2dae69"></script>
    <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="quipcell" href="api/modules.html" />
    <link rel="prev" title="quipcell" href="readme.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  

  
  

  </head><body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          

          <div class="body" role="main">
            
  <section id="Example-application-of-quipcell-to-human-lung-cell-atlas">
<h1>Example application of quipcell to human lung cell atlas<a class="headerlink" href="#Example-application-of-quipcell-to-human-lung-cell-atlas" title="Link to this heading">¶</a></h1>
<p>In this vignette, we demonstrate the application of <code class="docutils literal notranslate"><span class="pre">quipcell</span></code> to bulk deconvolution, replicating some analyses from the manuscript.</p>
<p>For our dataset, we use the Human Lung Cell Atlas, which consists of several studies. We hold out one of the studies as a test set, which we use to generate (pseudo)bulk samples with known ground truth.</p>
<p>To make this analysis runnable on a laptop, the data has been preprocessed into a smaller form (just the raw counts of the highly variable genes). See the <a class="reference external" href="https://github.com/Genentech/quipcell/blob/main/vignette/Readme.md">vignette readme</a> for instructions on how to download or generate the preprocessed data before running this notebook.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>from collections import Counter

import logging
logging.basicConfig(level=logging.INFO)

import numpy as np
import pandas as pd

import plotnine as gg
import mizani
import mizani.transforms as scales

import anndata as ad
import scanpy as sc

import scipy.sparse

from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.preprocessing import OneHotEncoder

import quipcell as qpc
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
WARNING:matplotlib.font_manager:Matplotlib is building the font cache; this may take a moment.
INFO:matplotlib.font_manager:Failed to extract font properties from /System/Library/Fonts/Supplemental/NISC18030.ttf: In FT2Font: Could not set the fontsize (invalid pixel size; error code 0x17)
INFO:matplotlib.font_manager:Failed to extract font properties from /System/Library/Fonts/LastResort.otf: tuple indices must be integers or slices, not str
INFO:matplotlib.font_manager:Failed to extract font properties from /System/Library/Fonts/Apple Color Emoji.ttc: In FT2Font: Could not set the fontsize (invalid pixel size; error code 0x17)
INFO:matplotlib.font_manager:generated new fontManager
</pre></div></div>
</div>
<section id="Load-single-cell-data">
<h2>Load single cell data<a class="headerlink" href="#Load-single-cell-data" title="Link to this heading">¶</a></h2>
<p>First, we load the (preprocessed) Human Lung Cell Atlas data.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata = ad.read_h5ad(&quot;data/hlca_hvgs.h5ad&quot;)
</pre></div>
</div>
</div>
<p>Next, we subset to 1/8th of the dataset to improve speed. We also exclude one of the studies as a validation test set to benchmark bulk deconvolution.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>validation_study = &#39;Krasnow_2020&#39;

rng = np.random.default_rng(12345)
n = adata.obs.shape[0]

keep = adata.obs[&#39;study&#39;] != validation_study
keep = keep &amp; (rng.random(size=n) &lt; .125)

adata_ref = adata[keep,:]
print(f&quot;{adata_ref.shape[0]} cells in reference set&quot;)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
65531 cells in reference set
</pre></div></div>
</div>
</section>
<section id="Normalize-and-transform-single-cell-data">
<h2>Normalize and transform single cell data<a class="headerlink" href="#Normalize-and-transform-single-cell-data" title="Link to this heading">¶</a></h2>
<p>Next, normalize the raw counts. The preprocessing script saved the total UMIs per cell before filtering to highly variable genes, so we divide it out, and normalize to counts per 1000 UMIs.</p>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># Normalize
normalize_sum = 1000
adata_ref.X = scipy.sparse.diags(normalize_sum / adata_ref.obs[&#39;n_umi&#39;].values) @ adata_ref.X
</pre></div>
</div>
</div>
<p>Next, we run PCA, and then perform Linear Discriminant Analysis on the PCs and celltype labels. This will be the embedding space we use for <code class="docutils literal notranslate"><span class="pre">quipcell</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>sc.pp.pca(adata_ref, n_comps=100)

# since the PCA is zero-centered, we need to save the gene-wise means
# to apply PCA rotation to pseudobulks
gene_means = adata_ref.X.mean(axis=0)
gene_means = np.squeeze(np.asarray(gene_means))
adata_ref.var[&#39;x_mean&#39;] = gene_means

lda = LinearDiscriminantAnalysis(n_components=15)

X = adata_ref.obsm[&#39;X_pca&#39;]
y = adata_ref.obs[&#39;ann_finest_level&#39;]

lda.fit(X, y)

adata_ref.obsm[&#39;X_lda&#39;] = lda.transform(adata_ref.obsm[&#39;X_pca&#39;])
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/kammj2/miniforge3/envs/quipcell/lib/python3.12/site-packages/scanpy/preprocessing/_pca.py:314: ImplicitModificationWarning: Setting element `.obsm[&#39;X_pca&#39;]` of view, initializing view as actual.
</pre></div></div>
</div>
</section>
<section id="Load-and-transform-(pseudo)bulk-data">
<h2>Load and transform (pseudo)bulk data<a class="headerlink" href="#Load-and-transform-(pseudo)bulk-data" title="Link to this heading">¶</a></h2>
<p>Next, we read in the pseudobulk counts (generated in <code class="docutils literal notranslate"><span class="pre">preprocessing.py</span></code>), and subset to the samples in the validation/test set.</p>
<p>We then normalize the counts, and subset to the same genes as the reference single cell set.</p>
<p>Then, we apply a linear transformation to map the bulk samples to the embedding space.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>adata_pseudobulk = ad.read_h5ad(&quot;data/pseudobulks.h5ad&quot;)

# Subset to validation set
keep = adata_pseudobulk.obs[&#39;study&#39;] == validation_study
adata_pseudobulk = adata_pseudobulk[keep,:]

# Normalize on the same scale as the single cell reference set
adata_pseudobulk.raw = adata_pseudobulk
sc.pp.normalize_total(adata_pseudobulk, target_sum=normalize_sum)

# Subset to same genes as the single cell reference set
adata_pseudobulk = adata_pseudobulk[:, adata.var.index]

# Apply PCA rotation to pseudobulks
X = adata_pseudobulk.X - adata_ref.var[&#39;x_mean&#39;].values
X = np.asarray(X @ adata_ref.varm[&#39;PCs&#39;])
adata_pseudobulk.obsm[&#39;X_pca&#39;] = X
# Apply LDA rotation to pseudobulks
adata_pseudobulk.obsm[&#39;X_lda&#39;] = lda.transform(X)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/var/folders/rw/hpynwn2500q6m7_3rdsw2zcc0000gp/T/ipykernel_23386/4126115140.py:17: ImplicitModificationWarning: Setting element `.obsm[&#39;X_pca&#39;]` of view, initializing view as actual.
</pre></div></div>
</div>
</section>
<section id="Estimate-weights-for-read-level-abundances">
<h2>Estimate weights for read-level abundances<a class="headerlink" href="#Estimate-weights-for-read-level-abundances" title="Link to this heading">¶</a></h2>
<p>Now that the single cell reference set and the bulk samples are in the same embedding space, we can estimate the weights with <code class="docutils literal notranslate"><span class="pre">quipcell</span></code>.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>w_umi = qpc.estimate_weights_multisample(adata_ref.obsm[&#39;X_lda&#39;],
                                         adata_pseudobulk.obsm[&#39;X_lda&#39;])
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:quipcell._solvers:Solving for sample i=0
INFO:quipcell._solvers:objective=2.208068696366285, optimal
INFO:quipcell._solvers:Solving for sample i=1
INFO:quipcell._solvers:objective=1.3962415069956384, optimal
INFO:quipcell._solvers:Solving for sample i=2
INFO:quipcell._solvers:objective=2.1952911740312, optimal
INFO:quipcell._solvers:Solving for sample i=3
INFO:quipcell._solvers:objective=1.071730756157023, optimal
INFO:quipcell._solvers:Solving for sample i=4
INFO:quipcell._solvers:objective=0.4401995928507547, optimal
</pre></div></div>
</div>
</section>
<section id="Convert-read-level-abundance-to-cell-level-abundance">
<h2>Convert read-level abundance to cell-level abundance<a class="headerlink" href="#Convert-read-level-abundance-to-cell-level-abundance" title="Link to this heading">¶</a></h2>
<p>Note that the gene counts of bulk samples don’t normalize for the fact that different cell types have different number of reads. Therefore, the weights represent a probability distribution of a random <em>read</em> drawn from the bulk sample, rather than a probability distribution over a random <em>cell</em> from the bulk sample.</p>
<p>We can convert the read-level weights to cell-level weights by estimating size factors for each of the reference cells, and normalizing the weights by those.</p>
<p>If the reference set doesn’t have too much technical variation, then we can just use the number of UMIs per cell as the normalizing factor. However, the HLCA is a highly heterogeneous dataset, consisting of many samples from different studies, sequenced to different depths, and containing different celltypes. Therefore, we need to control for sample when estimating the size factors. <code class="docutils literal notranslate"><span class="pre">quipcell</span></code> provides a method to estimate size factors while controlling for sample by using Poisson regression.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># Estimate size factors using Poisson regression
size_factors = qpc.estimate_size_factors(
    adata_ref.obsm[&#39;X_lda&#39;],
    adata_ref.obs[&#39;n_umi&#39;].values,
    adata_ref.obs[&#39;sample&#39;].values,
    #verbose=True
)

# Convert read-level weights to cell-level weights
w_cell = qpc.renormalize_weights(w_umi, size_factors)
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/kammj2/miniforge3/envs/quipcell/lib/python3.12/site-packages/sklearn/linear_model/_linear_loss.py:294: RuntimeWarning: invalid value encountered in matmul
/Users/kammj2/miniforge3/envs/quipcell/lib/python3.12/site-packages/sklearn/linear_model/_linear_loss.py:294: RuntimeWarning: invalid value encountered in matmul
</pre></div></div>
</div>
</section>
<section id="Plot-inferred-weights">
<h2>Plot inferred weights<a class="headerlink" href="#Plot-inferred-weights" title="Link to this heading">¶</a></h2>
<p>Below, we plot the read-level and cell-level weights on the reference UMAP for each of the 5 validation bulk samples.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span># Dataframe for plotting UMI-level weights on UMAP
df = pd.DataFrame(
    np.hstack([adata_ref.obsm[&#39;X_umap&#39;], w_umi]),
    columns = [&#39;UMAP1&#39;, &#39;UMAP2&#39;] + list(adata_pseudobulk.obs.index)
).melt( # pivot to long format
    id_vars=[&#39;UMAP1&#39;, &#39;UMAP2&#39;],
    var_name=&#39;sample&#39;,
    value_name=&#39;weight&#39;
)

# set very small weights to 0 for prettier plotting
weight_trunc = df[&#39;weight&#39;].values.copy()
weight_trunc[weight_trunc &lt; 1e-9] = 0
df[&#39;weight_trunc&#39;] = weight_trunc

(gg.ggplot(df, gg.aes(x=&quot;UMAP1&quot;, y=&quot;UMAP2&quot;, color=&quot;weight_trunc&quot;)) +
    gg.geom_point(size=.25, alpha=.5) +
    gg.facet_wrap(&quot;~sample&quot;) +
    gg.scale_color_cmap(trans=scales.log_trans(base=10)) +
    gg.ggtitle(&quot;Read-level weights&quot;))
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/kammj2/miniforge3/envs/quipcell/lib/python3.12/site-packages/pandas/core/arraylike.py:399: RuntimeWarning: divide by zero encountered in log10
INFO:matplotlib.font_manager:Fontsize 0.00 &lt; 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/vignette_17_1.png" class="no-scaled-link" src="_images/vignette_17_1.png" style="width: 640px; height: 480px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span> # Dataframe for plotting cell-level weights on UMAP
df = pd.DataFrame(
    np.hstack([adata_ref.obsm[&#39;X_umap&#39;], w_cell]),
    columns = [&#39;UMAP1&#39;, &#39;UMAP2&#39;] + list(adata_pseudobulk.obs.index)
).melt( # pivot to long format
    id_vars=[&#39;UMAP1&#39;, &#39;UMAP2&#39;],
    var_name=&#39;sample&#39;,
    value_name=&#39;weight&#39;
)

# set very small weights to 0 for prettier plotting
weight_trunc = df[&#39;weight&#39;].values.copy()
weight_trunc[weight_trunc &lt; 1e-9] = 0
df[&#39;weight_trunc&#39;] = weight_trunc

(gg.ggplot(df, gg.aes(x=&quot;UMAP1&quot;, y=&quot;UMAP2&quot;, color=&quot;weight_trunc&quot;)) +
    gg.geom_point(size=.25, alpha=.5) +
    gg.facet_wrap(&quot;~sample&quot;) +
    gg.scale_color_cmap(trans=scales.log_trans(base=10)) +
    gg.ggtitle(&quot;Cell-level weights&quot;))
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
/Users/kammj2/miniforge3/envs/quipcell/lib/python3.12/site-packages/pandas/core/arraylike.py:399: RuntimeWarning: divide by zero encountered in log10
INFO:matplotlib.font_manager:Fontsize 0.00 &lt; 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/vignette_18_1.png" class="no-scaled-link" src="_images/vignette_18_1.png" style="width: 640px; height: 480px;" />
</div>
</div>
</section>
<section id="Plot-true-vs-estimated-abundances">
<h2>Plot true vs estimated abundances<a class="headerlink" href="#Plot-true-vs-estimated-abundances" title="Link to this heading">¶</a></h2>
<p>Next, we compare the estimated weights to the true celltype proportions. In the preprocessing script, we precomputed the fraction of reads and cells coming from each celltype, at different levels of celltype annotation. We read in these precomputed abundances below.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>df_abundance = pd.read_csv(&#39;data/abundances.csv&#39;)
# filter to samples in the validation set
df_abundance = df_abundance[df_abundance[&#39;sample&#39;].isin(adata_pseudobulk.obs.index)]
# convert sample name to string to prevent missing categorical levels
df_abundance[&#39;sample&#39;] = df_abundance[&#39;sample&#39;].astype(str)
df_abundance
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sample</th>
      <th>ann_level</th>
      <th>celltype</th>
      <th>n_umi</th>
      <th>frac_umi</th>
      <th>n_cell</th>
      <th>frac_cell</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>644</th>
      <td>distal 1a</td>
      <td>lvl1</td>
      <td>Endothelial</td>
      <td>4461681</td>
      <td>0.060048</td>
      <td>1495</td>
      <td>0.198724</td>
    </tr>
    <tr>
      <th>645</th>
      <td>distal 1a</td>
      <td>lvl1</td>
      <td>Epithelial</td>
      <td>8394719</td>
      <td>0.112981</td>
      <td>853</td>
      <td>0.113386</td>
    </tr>
    <tr>
      <th>646</th>
      <td>distal 1a</td>
      <td>lvl1</td>
      <td>Immune</td>
      <td>59114660</td>
      <td>0.795600</td>
      <td>4719</td>
      <td>0.627276</td>
    </tr>
    <tr>
      <th>647</th>
      <td>distal 1a</td>
      <td>lvl1</td>
      <td>Stroma</td>
      <td>2330951</td>
      <td>0.031371</td>
      <td>456</td>
      <td>0.060614</td>
    </tr>
    <tr>
      <th>648</th>
      <td>distal 2</td>
      <td>lvl1</td>
      <td>Endothelial</td>
      <td>30078564</td>
      <td>0.205619</td>
      <td>8449</td>
      <td>0.455693</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>25559</th>
      <td>proximal 3</td>
      <td>lvl5</td>
      <td>Subpleural fibroblasts</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>25560</th>
      <td>proximal 3</td>
      <td>lvl5</td>
      <td>Suprabasal</td>
      <td>10395727</td>
      <td>0.130773</td>
      <td>482</td>
      <td>0.062041</td>
    </tr>
    <tr>
      <th>25561</th>
      <td>proximal 3</td>
      <td>lvl5</td>
      <td>T cells proliferating</td>
      <td>34048</td>
      <td>0.000428</td>
      <td>1</td>
      <td>0.000129</td>
    </tr>
    <tr>
      <th>25562</th>
      <td>proximal 3</td>
      <td>lvl5</td>
      <td>Tuft</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0</td>
      <td>0.000000</td>
    </tr>
    <tr>
      <th>25563</th>
      <td>proximal 3</td>
      <td>lvl5</td>
      <td>pre-TB secretory</td>
      <td>1892948</td>
      <td>0.023812</td>
      <td>144</td>
      <td>0.018535</td>
    </tr>
  </tbody>
</table>
<p>770 rows × 7 columns</p>
</div></div>
</div>
<p>Next, we compute estimated celltype abundances, by summing over the weights fit by <code class="docutils literal notranslate"><span class="pre">quipcell</span></code>. We add the estimated number of reads and cells coming from each celltype to the dataframe of abundances.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>est_frac_umi = []
est_frac_cells = []

# TODO use a more descriptive variable name than i
for i in range(1, 6):
    enc = OneHotEncoder()
    mat_onehot = enc.fit_transform(
        adata_ref.obs[f&#39;celltype_lvl{i}&#39;].values.reshape(-1,1)
    )

    # Function to sum weights by celltype, then pivot to long dataframe
    def aggregate_and_reshape(w):
        return pd.DataFrame(
            w.T @ mat_onehot,
            columns = enc.categories_[0],
            index = adata_pseudobulk.obs.index
        ).reset_index(names=&#39;sample&#39;).melt(
            id_vars=[&#39;sample&#39;], var_name=&#39;celltype&#39;, value_name=&#39;weight&#39;
        ).assign(ann_level=f&#39;lvl{i}&#39;)

    est_frac_umi.append(aggregate_and_reshape(w_umi))
    est_frac_cells.append(aggregate_and_reshape(w_cell))


# Function to combine the summed weights across annotation levels,
# returning a vector to be added to df_abundance
def concat_aggregated_weights(agg_weight_list):
    idx_keys = [&#39;ann_level&#39;, &#39;sample&#39;, &#39;celltype&#39;]
    return (pd.concat(agg_weight_list)
            .set_index(idx_keys)
            .loc[zip(*[df_abundance[k] for k in idx_keys])]
            .values)

df_abundance[&#39;est_frac_umi&#39;] = concat_aggregated_weights(est_frac_umi)
df_abundance[&#39;est_frac_cell&#39;] = concat_aggregated_weights(est_frac_cells)

df_abundance
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>sample</th>
      <th>ann_level</th>
      <th>celltype</th>
      <th>n_umi</th>
      <th>frac_umi</th>
      <th>n_cell</th>
      <th>frac_cell</th>
      <th>est_frac_umi</th>
      <th>est_frac_cell</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>644</th>
      <td>distal 1a</td>
      <td>lvl1</td>
      <td>Endothelial</td>
      <td>4461681</td>
      <td>0.060048</td>
      <td>1495</td>
      <td>0.198724</td>
      <td>0.055587</td>
      <td>0.184202</td>
    </tr>
    <tr>
      <th>645</th>
      <td>distal 1a</td>
      <td>lvl1</td>
      <td>Epithelial</td>
      <td>8394719</td>
      <td>0.112981</td>
      <td>853</td>
      <td>0.113386</td>
      <td>0.085862</td>
      <td>0.085786</td>
    </tr>
    <tr>
      <th>646</th>
      <td>distal 1a</td>
      <td>lvl1</td>
      <td>Immune</td>
      <td>59114660</td>
      <td>0.795600</td>
      <td>4719</td>
      <td>0.627276</td>
      <td>0.843633</td>
      <td>0.710039</td>
    </tr>
    <tr>
      <th>647</th>
      <td>distal 1a</td>
      <td>lvl1</td>
      <td>Stroma</td>
      <td>2330951</td>
      <td>0.031371</td>
      <td>456</td>
      <td>0.060614</td>
      <td>0.014919</td>
      <td>0.019973</td>
    </tr>
    <tr>
      <th>648</th>
      <td>distal 2</td>
      <td>lvl1</td>
      <td>Endothelial</td>
      <td>30078564</td>
      <td>0.205619</td>
      <td>8449</td>
      <td>0.455693</td>
      <td>0.177143</td>
      <td>0.379683</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>25559</th>
      <td>proximal 3</td>
      <td>lvl5</td>
      <td>Subpleural fibroblasts</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.000464</td>
      <td>0.000488</td>
    </tr>
    <tr>
      <th>25560</th>
      <td>proximal 3</td>
      <td>lvl5</td>
      <td>Suprabasal</td>
      <td>10395727</td>
      <td>0.130773</td>
      <td>482</td>
      <td>0.062041</td>
      <td>0.036868</td>
      <td>0.029626</td>
    </tr>
    <tr>
      <th>25561</th>
      <td>proximal 3</td>
      <td>lvl5</td>
      <td>T cells proliferating</td>
      <td>34048</td>
      <td>0.000428</td>
      <td>1</td>
      <td>0.000129</td>
      <td>0.000636</td>
      <td>0.001184</td>
    </tr>
    <tr>
      <th>25562</th>
      <td>proximal 3</td>
      <td>lvl5</td>
      <td>Tuft</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0</td>
      <td>0.000000</td>
      <td>0.000277</td>
      <td>0.000255</td>
    </tr>
    <tr>
      <th>25563</th>
      <td>proximal 3</td>
      <td>lvl5</td>
      <td>pre-TB secretory</td>
      <td>1892948</td>
      <td>0.023812</td>
      <td>144</td>
      <td>0.018535</td>
      <td>0.007603</td>
      <td>0.006743</td>
    </tr>
  </tbody>
</table>
<p>770 rows × 9 columns</p>
</div></div>
</div>
<p>Finally, we plot the estimated vs actual abundances, at both the read and cell level.</p>
<p>At the finest resolution (annotation level 5), there is some inaccuracy because a common celltype in the validation set (CCL3+ Alveolar macrophages) is very rare in the reference samples, and cannot be distinguished from other Alveolar macrophages in the embedding space. See the manuscript for more details.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span> (gg.ggplot(df_abundance.sample(frac=1, random_state=42),
           gg.aes(x=&quot;frac_umi&quot;, y=&quot;est_frac_umi&quot;, color=&quot;ann_level&quot;, shape=&quot;sample&quot;)) +
    gg.geom_point() +
    gg.geom_abline(linetype=&#39;dashed&#39;) +
    gg.scale_x_sqrt(breaks=[.01,.1,.2,.4,.6,.8]) +
    gg.scale_y_sqrt(breaks=[.01,.1,.2,.4,.6,.8]) +
    gg.theme_bw(base_size=16) +
    gg.ggtitle(&quot;Read-level accuracy&quot;))
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:matplotlib.font_manager:Fontsize 0.00 &lt; 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/vignette_24_1.png" class="no-scaled-link" src="_images/vignette_24_1.png" style="width: 640px; height: 480px;" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span> (gg.ggplot(df_abundance.sample(frac=1, random_state=42),
           gg.aes(x=&quot;frac_cell&quot;, y=&quot;est_frac_cell&quot;, color=&quot;ann_level&quot;, shape=&quot;sample&quot;)) +
    gg.geom_point() +
    gg.geom_abline(linetype=&#39;dashed&#39;) +
    gg.scale_x_sqrt(breaks=[.01,.1,.2,.4,.6,.8]) +
    gg.scale_y_sqrt(breaks=[.01,.1,.2,.4,.6,.8]) +
    gg.theme_bw(base_size=16) +
    gg.ggtitle(&quot;Cell-level accuracy&quot;))
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:matplotlib.font_manager:Fontsize 0.00 &lt; 1.0 pt not allowed by FreeType. Setting fontsize = 1 pt
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/vignette_25_1.png" class="no-scaled-link" src="_images/vignette_25_1.png" style="width: 640px; height: 480px;" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</section>
</section>


          </div>
          
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
<h1 class="logo"><a href="index.html">quipcell</a></h1>








<h3>Navigation</h3>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="readme.html">Overview</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Vignette</a></li>
<li class="toctree-l1"><a class="reference internal" href="api/modules.html">Module Reference</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributions &amp; Help</a></li>
<li class="toctree-l1"><a class="reference internal" href="license.html">License</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Authors</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
</ul>

<div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
      <li>Previous: <a href="readme.html" title="previous chapter">quipcell</a></li>
      <li>Next: <a href="api/modules.html" title="next chapter">quipcell</a></li>
  </ul></li>
</ul>
</div>
<search id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" autocomplete="off" autocorrect="off" autocapitalize="off" spellcheck="false"/>
      <input type="submit" value="Go" />
    </form>
    </div>
</search>
<script>document.getElementById('searchbox').style.display = "block"</script>








        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &#169;2024, Jack Kamm.
      
      |
      Powered by <a href="https://www.sphinx-doc.org/">Sphinx 7.3.7</a>
      &amp; <a href="https://alabaster.readthedocs.io">Alabaster 0.7.16</a>
      
      |
      <a href="_sources/vignette.ipynb.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>